From: Mykyta Yatsenko <yatsenko@meta.com>
Subject: maccess: fix strncpy_from_user_nofault empty string handling
Date: Thu, 17 Apr 2025 16:28:08 +0100

strncpy_from_user_nofault() should return the length of the copied string
including the trailing NUL, but if the argument unsafe_addr points to an
empty string ({'\0'}), the return value is 0.

This happens as strncpy_from_user() copies terminal symbol into dst and
returns 0 (as expected), but strncpy_from_user_nofault does not modify ret
as it is not equal to count and not greater than 0, so 0 is returned,
which contradicts the contract.

Link: https://lkml.kernel.org/r/20250417152808.722409-1-mykyta.yatsenko5@gmail.com
Signed-off-by: Mykyta Yatsenko <yatsenko@meta.com>
Cc: Andrii Nakryiko <andrii@kernel.org>
Cc: "Masami Hiramatsu (Google)" <mhiramat@kernel.org>
Cc: Steven Rostedt <rostedt@goodmis.org>
Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
---

 kernel/trace/trace_events_filter.c |   10 ++++++++--
 mm/maccess.c                       |    2 +-
 2 files changed, 9 insertions(+), 3 deletions(-)

--- a/kernel/trace/trace_events_filter.c~maccess-fix-strncpy_from_user_nofault-empty-string-handling
+++ a/kernel/trace/trace_events_filter.c
@@ -800,6 +800,7 @@ static __always_inline char *test_string
 {
 	struct ustring_buffer *ubuf;
 	char *kstr;
+	int cnt;
 
 	if (!ustring_per_cpu)
 		return NULL;
@@ -808,7 +809,9 @@ static __always_inline char *test_string
 	kstr = ubuf->buffer;
 
 	/* For safety, do not trust the string pointer */
-	if (!strncpy_from_kernel_nofault(kstr, str, USTRING_BUF_SIZE))
+	cnt = strncpy_from_kernel_nofault(kstr, str, USTRING_BUF_SIZE);
+	/* Return null if empty string or error */
+	if (cnt <= 1)
 		return NULL;
 	return kstr;
 }
@@ -818,6 +821,7 @@ static __always_inline char *test_ustrin
 	struct ustring_buffer *ubuf;
 	char __user *ustr;
 	char *kstr;
+	int cnt;
 
 	if (!ustring_per_cpu)
 		return NULL;
@@ -827,7 +831,9 @@ static __always_inline char *test_ustrin
 
 	/* user space address? */
 	ustr = (char __user *)str;
-	if (!strncpy_from_user_nofault(kstr, ustr, USTRING_BUF_SIZE))
+	cnt = strncpy_from_user_nofault(kstr, ustr, USTRING_BUF_SIZE);
+	/* Return null if empty string or error */
+	if (cnt <= 1)
 		return NULL;
 
 	return kstr;
--- a/mm/maccess.c~maccess-fix-strncpy_from_user_nofault-empty-string-handling
+++ a/mm/maccess.c
@@ -196,7 +196,7 @@ long strncpy_from_user_nofault(char *dst
 	if (ret >= count) {
 		ret = count;
 		dst[ret - 1] = '\0';
-	} else if (ret > 0) {
+	} else if (ret >= 0) {
 		ret++;
 	}
 
_
